Added feature in Latex template to censor text wrapped with double ba…

kpence opened the pull request:
…rs (e.g. ||censored text||). This is easy to cancel by adding \StopCensoring command at the beginning of the latex document.

kpence committed:
Added feature in Latex template to censor text wrapped with double bars (e.g. ||censored text||). This is easy to cancel by adding \StopCensoring command at the beginning of the latex document.

karimElmougi commented on test_resume_censored.yaml:
```diff
@@ -0,0 +1,79 @@
+header:
+  name: '||Karim Elmougi||'
```
It is unfortunate to have to resort to quotes due to `|` clashing with YAML. Is there perhaps another symbol that could be used?

kpence replied:
I'll look into what could be done later.

karimElmougi commented on test_resume_censored.yaml:
```diff
@@ -0,0 +1,79 @@
+header:
```
I think I would just add some censoring bits to `test_resume.yaml` rather than creating a mostly identical copy.

kpence replied:
So what I did in the last commit was just delete the identical copy.

karimElmougi commented on internal/resume/templates/latex.go:
```diff
@@ -25,6 +26,12 @@ func latexEscape(s string) string {
 }
 
+func latexCensor(s string) string {
+  re := regexp.MustCompile(`\|\|(.*)\|\|`)
```
You have some inconsistent whitespace (spaces vs tabs) here. Make sure to run `gotfmt` on the project. It's Go's standard formatting tool.

kpence replied:
Thanks! Done.

karimElmougi commented on internal/resume/templates/latex.go:
```diff
@@ -25,6 +26,12 @@ func latexEscape(s string) string {
 }
 
+func latexCensor(s string) string {
+  re := regexp.MustCompile(`\|\|(.*)\|\|`)
```
I don't use them, but would this regex hypothetically work on multiline strings?

kpence replied:
I just ran some tests by adding multiline strings to my resume YAML, but the rendered output seems to ignore the newlines.
Otherwise there weren't any issues I found with the censoring, it acted as if there were no newlines basically.

karimElmougi commented on internal/resume/templates/latex.go:
```diff
@@ -38,6 +45,7 @@ var latexDocument = `
+    \usepackage{censor}
 
 %%%%%%%%%%%%%%%%%%%%%%% DEFINITIONS FOR RESUME %%%%%%%%%%%%%%%%%%%%%%%
 
```
I think the `\StopCensoring` command should be added in the Latex template. Adding a `CensoringEnabled` boolean flag in the `Resume` struct in `internal/resume` seems like a good way to do it. Then in the template you could have `[[ if not .CensoringEnabled ]]\StopCensoring[[end -]]`. 

Doing it that way, it would be best to add censoring to the plaintext template as well, for consistency.

kpence replied:
I implemented this in the latest commit.

kpence committed:
I changed the regex to use a lazy quantifier so that now you can have multiple sections of text censored out.

kpence committed:
Added censoring flag, and plaintext censoring.

Fixed indentation and formatting with gofmt. Censoring can be configured
on or off in the YAML schema by setting the `censoringenabled: true/false` key value.
Plaintext now has censoring, it replaces the characters of censored text with '#'
characters.

kpence commented:
Check out the new commit please :)

karimElmougi commented on internal/resume/templates/plaintext.go:
```diff
@@ -1,47 +1,59 @@
 
+func plaintextCensor(s string) string {
+  re := regexp.MustCompile(`\|\|.*?\|\|`)
+  //func test(_s string) { return strings.Repeat("#",len(_s)-4) })
```
Remove this commented out code. And don't forget to run `gofmt`!

kpence replied:
Woops.

I thought I `gofmt -w path/to/code.go`. Maybe I should automate it so that gofmt gets run before committing....

karimElmougi replied:
`gofmt -s -w .` from the root directory ought to take care of it. That or `go fmt ./...`

kpence committed:
Removed comment and reformated

karimElmougi commented on internal/resume/templates/plaintext.go:
```diff
@@ -1,47 +1,59 @@
+func plaintextCensor(s string) string {
+  re := regexp.MustCompile(`\|\|.*?\|\|`)
+  //func test(_s string) { return strings.Repeat("#",len(_s)-4) })
+	s = re.ReplaceAllStringFunc(s, func(_s string) string { return strings.Repeat("#",len(_s)-4) })
```
In Go, underscores are generally used to signal discarded variables/parameters. Perhaps renaming `s` to `text` and `_s` to `s` would be cleaner? Something similar would need to be done in the latex template's censor function, for consistency.

karimElmougi commented:
It would be important to update the unit tests for the two templates, as they currently fail. The tests can be run with `go test ./...` from the repo's root directory.

kpence commented:
I'm confused by the tests, I don't know how they were supposed to work originally.

The error I'm getting is this:

```
➜  resumake git:(master) ✗ go test ./internal/resume/templates/latex_test.go
# command-line-arguments_test [command-line-arguments.test]
internal/resume/templates/latex_test.go:19:25: undefined: testResume
FAIL    command-line-arguments [build failed]
FAIL
```

I don't know where testResume is supposed to be defined. I think I'll work on this tomorrow.

karimElmougi commented:
Just run the tests with `go test ./...` instead of a specific file.

karimElmougi commented on internal/resume/resume.go:
```diff
@@ -7,6 +7,7 @@ import (
 )
 
 type Resume struct {
+	CensoringEnabled bool             `yaml:"censoringenabled"`
```
This should be checked in `resume_test.go`

karimElmougi replied:
And the yaml variable should follow camelCase, so `censoringEnabled`

kpence replied:
Done

karimElmougi commented:
To properly turn off censoring in the plaintext version, you could proceed like so:

1. Add a `censoringEnabled` global variable in `plaintext.go`
2. Change `plaintextCensor()` to 
 ```go
if censoringEnabled {
    re := regexp.MustCompile(`\|\|.*?\|\|`)
    s = re.ReplaceAllStringFunc(s, func(_s string) string { return strings.Repeat("#", len(_s)-4) })
} else {
    s = strings.ReplaceAll(s, "||", "")
}
return s
```
3. Add a `setCensoringEnabled()` function like so:
```go
func setCensoringEnabled(b bool) string {
    censoringEnabled = b
    return ""
}
```
4. Put `{{- .CensoringEnabled | setCensoringEnabled -}}` at the top of the plaintext template

An even cleaner way to do this would be to use lambdas to capture a heap-allocated `bool` rather than using a package variable like this, but the idea remains the same.

kpence committed:
Fixed censoringenabled in plaintext & fixed tests

Now plaintext template will behave correctly according to whether censoringenabled
is set to true or false.
The argument s is changed to text.
The tests are now working, which required a change in latex.go:68.
resume_test.go now checks for CensoringEnabled as it should.

karimElmougi commented on internal/resume/resume_test.go:
```diff
@@ -21,6 +21,7 @@ func TestResumeDeserialization(t *testing.T) {
 
 	expectedTime := time.Date(2004, 01, 1, 0, 0, 0, 0, time.UTC)
 	expected := resume.Resume{
+		CensoringEnabled: false,
```
Gonna need to check the value as well, with `g.Expect(r.CensoringEnabled).To(Equal(expected.CensoringEnabled))`

kpence replied:
Woops

kpence committed:
Added value check for CensoringEnabled to resume_test.

kpence commented:
oh no I forgot to use gofmt, sorry.

kpence commented:
Nevermind, I didn't need to use gofmt, I was confused because of how the code appeared on github.

karimElmougi commented:
Just need to rename the `censoringenabled` yaml variable to `censoringEnabled`, as camelCase is idiomatic in yaml.

kpence committed:
Added missing vspace in skills section to latex template

karimElmougi commented on internal/resume/templates/latex.go:
```diff
@@ -111,6 +111,7 @@ var latexDocument = `
 
 %==== Skills ====%
 \header{Skills}
+\vspace{1mm}
```
Might as well add this to the project section as well.

kpence replied:
sorry i forgot to add it to my tests. I need to learn to use automation before pushing.

kpence committed:
Added missing vspace in projects section, updated latex test

kpence commented:
I think the added vspaces make it look better

kpence committed:
Renamed censoringenabled yaml variable to censoringEnabled (camelCase)

Because camelCase is idiomatic of yaml

karimElmougi reviewed (approved):
LGTM
